wildmat.c

```diff
@@ -0,0 +1,165 @@
+ * 
+ * Written by Rich $alz, mirror!rs, Wed Nov 26 19:03:17 EST 1986.
+ * Rich $alz is now <rsalz@osf.org>.
+ * April, 1991:  Replaced mutually-recursive calls with in-line code
+ * for the star character.
+ *
+ * Special thanks to Lars Mathiesen <thorinn@diku.dk> for the ABORT code.
+ * This can greatly speed up failing wildcard patterns.  For example:
+ * pattern: -*-*-*-*-*-*-12-*-*-*-m-*-*-*
+ * text 1:  -adobe-courier-bold-o-normal--12-120-75-75-m-70-iso8859-1
+ * text 2:  -adobe-courier-bold-o-normal--12-120-75-75-X-70-iso8859-1
+ * Text 1 matches with 51 calls, while text 2 fails with 54 calls.  Without
+ * the ABORT code, it takes 22310 calls to fail.  Ugh.  The following
+ * explanation is from Lars:
+ * The precondition that must be fulfilled is that DoMatch will consume
+ * at least one character in text.  This is true if *p is neither '*' no
+ * '\0'.)  The last return has ABORT instead of FALSE to avoid quadratic
+ * behaviour in cases like pattern "*a*b*c*d" with text "abcxxxxx".  With
+ * FALSE, each star-loop has to run to the end of the text; with ABORT
+ * only the last one does.
+ * 
+ * Once the control of one instance of DoMatch enters the star-loop, that
+ * instance will return either TRUE or ABORT, and any calling instance
+ * will therefore return immediately after (without calling recursively
+ * again).  In effect, only one star-loop is ever active.  It would be
+ * possible to modify the code to maintain this context explicitly,
+ * eliminating all recursive calls at the cost of some complication and
+ * loss of clarity (and the ABORT stuff seems to be unclear enough by
+ * itself).  I think it would be unwise to try to get this into a
+ * released version unless you have a good test data base to try it out
+ * on.
+ */
+#define TRUE			 1
+#define FALSE			 0
+#define ABORT			-1
+
+
+/*- What character marks an inverted character class?  */
+#define NEGATE_CLASS		'^'
+/*- Is "*" a common pattern?  */
+#define OPTIMIZE_JUST_STAR
+/*- Do tar(1) matching rules, which ignore a trailing slash?  */
+#undef MATCH_TAR_PATTERN
+
+
+/*- Match text and p, return TRUE, FALSE, or ABORT.  */
+static int
+DoMatch(text, p)
+	register char  *text;
+	register char  *p;
+{
+	register int    last;
+	register int    matched;
+	register int    reverse;
+
+	for (; *p; text++, p++)
+	{
+		if (*text == '\0' && *p != '*')
+			return ABORT;
+		switch (*p)
+		{
+		case '\\': /*- Literal match with following character.  */
+			p++;
+			/*- FALLTHROUGH */
+		default:
+			if (*text != *p)
+				return FALSE;
+			continue;
+		case '?': /*- Match anything. */
+			continue;
+		case '*':
+			/*- Consecutive stars act just like one.  */
+			while (*++p == '*')
+				continue;
+			/*- Trailing star matches everything.  */
+			if (*p == '\0')
+				return TRUE;
+			while (*text)
+				if ((matched = DoMatch(text++, p)) != FALSE)
+					return matched;
+			return ABORT;
+		case '[':
+			reverse = p[1] == NEGATE_CLASS ? TRUE : FALSE;
```

Comment:
I think that will overflow the input string if passed something that ends in "[" or "[^" as it will go one char too far when entering the for loop below.

Reply:
Indeed yes. I don't like the function now. Is there a better way? spawn-filter already has code to use regexec() if env variable QREGEX is set. I can remove wildmat.c totally and use regexec() instead. Or is there a better function you are aware of that can be used for pattern matching?

Reply:
Just thinking out aloud. Is the need for pattern matching for domains really required. regex matching or using wildmat can be totally avoided if we use exact domains in the control file filterargs.